File: implementations/java/vanilla/src/main/java/org/cwi/waebric/parser/ModuleParser.java
package org.cwi.waebric.parser;

import java.util.List;

import org.cwi.waebric.WaebricKeyword;
import org.cwi.waebric.WaebricSymbol;
import org.cwi.waebric.parser.ast.basic.IdCon;
import org.cwi.waebric.parser.ast.module.Import;
import org.cwi.waebric.parser.ast.module.Module;
import org.cwi.waebric.parser.ast.module.ModuleId;
import org.cwi.waebric.parser.ast.module.Modules;
import org.cwi.waebric.parser.ast.module.function.FunctionDef;
import org.cwi.waebric.parser.ast.module.site.Site;
import org.cwi.waebric.scanner.token.TokenIterator;
import org.cwi.waebric.scanner.token.WaebricTokenSort;

/**
 * module languages/waebric/syntax/Modules
 * 
 * @author Jeroen van Schagen
 * @date 20-05-2009
 */
class ModuleParser extends AbstractParser {

	private final SiteParser siteParser;
	private final FunctionParser functionParser;
	
	/**
	 * Construct module parser
	 * @param tokens
	 * @param exceptions
	 */
	public ModuleParser(TokenIterator tokens, List<SyntaxException> exceptions) {
		super(tokens, exceptions);
		
		// Initialize sub-parsers
		siteParser = new SiteParser(tokens, exceptions);
		functionParser = new FunctionParser(tokens, exceptions);
	}
	
	/**
	 * Module* -> Modules
	 * @throws SyntaxException 
	 */
	public Modules parseModules() throws SyntaxException  {
		Modules modules = new Modules();
		
		// Parse Module*
		while(tokens.hasNext()) {
			next(WaebricKeyword.MODULE, "Module", "\"Module\" ModuleId");
			Module module = parseModule();
			modules.add(module);
		}

		return modules;
	}
	

	/**
	 * "module" ModuleId ModuleElement* -> Module
	 * @throws SyntaxException
	 */
	public Module parseModule() throws SyntaxException {
		current(WaebricKeyword.MODULE, "Module", "\"Module\" ModuleId");
		
		Module module = new Module();
		module.setIdentifier(parseModuleId());
		
		// ModuleElement*
		while(tokens.hasNext()) {
			if(tokens.peek(1).getLexeme().equals(WaebricKeyword.MODULE)) { 
				break; // Break current module parse, as new module is detected
			}
			
			tokens.next(); // Set current to first module element
			if(tokens.current().getLexeme() == WaebricKeyword.IMPORT) {
				Import imprt = parseImport();
				module.addImport(imprt);
			} else if(tokens.current().getLexeme() == WaebricKeyword.SITE) {
				Site site = siteParser.parseSite();
				module.addSite(site);
			} else if(tokens.current().getLexeme() == WaebricKeyword.DEF) {
				FunctionDef def = functionParser.parseFunctionDef();
				module.addFunctionDef(def);
			} else {
				reportUnexpectedToken(tokens.current(), 
					"Module element", "\"import\", \"site\" or \"def\"");
			}
		}
		
		return module;
	}
	
	/**
	 * { IdCon "." }+ -> ModuleId
	 * @throws SyntaxException 
	 */
	public ModuleId parseModuleId() throws SyntaxException {
		ModuleId moduleId = new ModuleId();
		
		do {
			next(WaebricTokenSort.IDCON, "Module identifier", "head:[azAZ] body:[azAZ09]");
			moduleId.add(new IdCon(tokens.current()));
			
			// Parse potential separator
			if(tokens.hasNext() && tokens.peek(1).getLexeme().equals(WaebricSymbol.PERIOD)) {
				tokens.next(); // Accept "." and set current to next identifier
			} else {
				break; // No period detected, end of identifier
			}
		} while(tokens.hasNext());
		
		return moduleId;
	}
	
	/**
	 * "import ModuleId -> Import
	 * @throws SyntaxException 
	 */
	public Import parseImport() throws SyntaxException {
		Import imprt = new Import();
		imprt.setIdentifier(parseModuleId());
		return imprt;
	}

}

File: implementations/csharp/vanilla/parser/ModuleParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Lexer.Tokenizer;
using Parser.Ast.Module;
using Parser.Exceptions;

namespace Parser
{
    /// <summary>
    /// Parser which parses a module
    /// </summary>
    public class ModuleParser : AbstractParser
    {
        #region Private Members

        private SiteParser siteParser;
        private FunctionParser functionParser;

        #endregion

        #region Public Methods

        /// <summary>
        /// Create a ModuleParser
        /// </summary>
        /// <param name="tokenStream"></param>
        /// <param name="exceptionList"></param>
        public ModuleParser(TokenIterator tokenStream) : base(tokenStream)
        {
            //Create parsers for sub elements in a module
            siteParser = new SiteParser(tokenStream);
            functionParser = new FunctionParser(tokenStream);
        }

        /// <summary>
        /// Parse one module
        /// </summary>
        /// <param name="module">Module to parse</param>
        public Module ParseModule()
        {
            Module module = new Module();

            //Skip module token
            NextToken("module", "module identifier", "module");

            //Parse first the identifier of the module and set it
            ModuleId moduleIdentifier = ParseModuleId();
            module.SetModuleId(moduleIdentifier);
            
            //Look for elements like SITE, DEF, etc
            while (TokenStream.HasNext())
            {
                if (MatchValue(TokenStream.Peek(1).GetValue().ToString(), Waebric.WaebricKeyword.MODULE.ToString()))
                {   //New module found, raise exception, only one module per file allowed
                    throw new UnexpectedToken("Unexpected token, found second module definition:", "module", TokenStream.Peek(1).GetLine());
                }
                CurrentToken = TokenStream.NextToken();
                

                //Check for different elements which may appear in a module
                if(MatchValue(CurrentToken.GetValue().ToString(), Waebric.WaebricKeyword.DEF.ToString()))
                {   //Function definition found
                    module.AddFunctionDefinition(functionParser.ParseFunctionDefinition());
                }
                else if(MatchValue(CurrentToken.GetValue().ToString(), Waebric.WaebricKeyword.SITE.ToString()))
                {   //Site definition found, call siteparser
                    module.AddSite(siteParser.ParseSite());
                }
                else if(MatchValue(CurrentToken.GetValue().ToString(), Waebric.WaebricKeyword.IMPORT.ToString()))
                {   //Imports found
                    module.AddImport(ParseImport());
                }
                else
                {
                    //Exception
                    throw new UnexpectedToken("Unexpected token found:", CurrentToken.GetValue().ToString(), CurrentToken.GetLine());
                }
            }

            return module;
        }

        /// <summary>
        /// Parse a module identifier
        /// </summary>
        /// <param name="identifier">Identifier to parse</param>
        public ModuleId ParseModuleId()
        {
            ModuleId moduleId = new ModuleId();
            
            //parse module identifier
            while (TokenStream.HasNext())
            {
                if (NextToken("identifier", "module identifier.identifier", TokenType.IDENTIFIER))
                {
                    moduleId.AddIdentifier(CurrentToken.GetValue().ToString());
                }
                else
                {
                    //Raise exception
                    throw new UnexpectedToken("Unexpected token found:", CurrentToken.GetValue().ToString(), CurrentToken.GetLine());
                }

                if (TokenStream.HasNext() && TokenStream.Peek(1).GetValue().ToString() == ".")
                {   //Period, so another identifier will appear after this one
                    NextToken(".", "module identifier.identifier", '.');
                }
                else
                {
                    break; //No more module identifier stuff will appear
                }
            }

            return moduleId;
        }

        /// <summary>
        /// Parse import
        /// </summary>
        /// <returns></returns>
        public Import ParseImport()
        {
            Import import = new Import();
            if (TokenStream.HasNext() && TokenStream.Peek(1).GetType() == TokenType.IDENTIFIER)
            {
                import.SetModuleId(ParseModuleId());
            }
            else
            {
                //Raise exception
                throw new UnexpectedToken("Unexpected token found:", CurrentToken.GetValue().ToString(), CurrentToken.GetLine());
            }
            return import;
        }

        #endregion

    }
}
